File: frontend/src/types/pilot.ts
/** TypeScript interfaces for the pilot review dashboard. */

export interface PilotRecordSummary {
  record_id: string;
  description: string;
  title: string | null;
  award_count: number;
  proposal_count: number;
  document_count: number;
  duplicate_count: number;
  total_budget: number;
  has_title_gaps: boolean;
  has_award: boolean;
}

export interface PilotDashboard {
  records: PilotRecordSummary[];
  total_awards: number;
  total_proposals: number;
  total_documents: number;
  total_duplicates: number;
  total_budget: number;
}

export interface AwardRead {
  award_id: string;
  award_number: string | null;
  award_title: string | null;
  proposal_id: string | null;
  project_id: string | null;
  sponsor_organization_id: string | null;
  award_status: string | null;
  original_start_date: string | null;
  current_end_date: string | null;
  adjusted_budget: number | null;
  cumulative_funding_amount: number | null;
  needs_veras_title: boolean;
  title_source_type: string | null;
}

export interface ProposalRead {
  proposal_id: string;
  proposal_number: string | null;
  proposal_title: string | null;
  project_id: string | null;
  sponsor_organization_id: string | null;
  proposal_status: string | null;
  start_date: string | null;
  end_date: string | null;
  needs_veras_enrichment: boolean;
  title_source_type: string | null;
}

export interface ProjectRead {
  project_id: string;
  project_title: string | null;
  project_status: string | null;
  start_date: string | null;
  end_date: string | null;
  needs_veras_title: boolean;
}

export interface DocumentRead {
  document_id: string;
  document_type_code: string | null;
  document_type_label: string | null;
  parent_entity_type: string | null;
  parent_entity_id: string | null;
  linked_proposal_id: string | null;
  top_level_bucket: string | null;
  filename: string | null;
  extension: string | null;
  mime_type: string | null;
  file_size_bytes: number | null;
  storage_relative_path: string | null;
  is_content_duplicate: boolean;
  needs_review: string | null;
}

export interface PersonnelRead {
  personnel_id: string;
  full_name: string | null;
  last_name: string | null;
  first_name: string | null;
  person_type: string | null;
  department_organization_id: string | null;
  needs_identity_review: boolean;
}

export interface PilotRecordDetail {
  record_id: string;
  description: string;
  awards: AwardRead[];
  proposals: ProposalRead[];
  projects: ProjectRead[];
  documents: DocumentRead[];
  personnel: PersonnelRead[];
  document_count: number;
  duplicate_count: number;
  category_counts: Record<string, number>;
}

export interface DocumentFilterResult {
  items: DocumentRead[];
  total_count: number;
  page: number;
  page_size: number;
  category_counts: Record<string, number>;
}

export interface AllowedValue {
  allowed_value_id: string;
  group_name: string | null;
  code: string | null;
  label: string | null;
  description: string | null;
}

File: backend/app/schemas/pilot.py
"""Pydantic schemas for the pilot review dashboard."""

from __future__ import annotations

import datetime

from pydantic import BaseModel


# ---------------------------------------------------------------------------
# Entity read schemas
# ---------------------------------------------------------------------------

class OrganizationRead(BaseModel):
    organization_id: str
    organization_name: str | None = None
    organization_type: str | None = None
    source_system: str | None = None
    needs_lookup: bool = False
    model_config = {"from_attributes": True}


class PersonnelRead(BaseModel):
    personnel_id: str
    full_name: str | None = None
    last_name: str | None = None
    first_name: str | None = None
    person_type: str | None = None
    department_organization_id: str | None = None
    needs_identity_review: bool = False
    model_config = {"from_attributes": True}


class ProjectRead(BaseModel):
    project_id: str
    project_title: str | None = None
    project_status: str | None = None
    start_date: datetime.date | None = None
    end_date: datetime.date | None = None
    needs_veras_title: bool = False
    model_config = {"from_attributes": True}


class ProposalRead(BaseModel):
    proposal_id: str
    proposal_number: str | None = None
    proposal_title: str | None = None
    project_id: str | None = None
    sponsor_organization_id: str | None = None
    proposal_status: str | None = None
    start_date: datetime.date | None = None
    end_date: datetime.date | None = None
    needs_veras_enrichment: bool = False
    title_source_type: str | None = None
    model_config = {"from_attributes": True}


class AwardRead(BaseModel):
    award_id: str
    award_number: str | None = None
    award_title: str | None = None
    proposal_id: str | None = None
    project_id: str | None = None
    sponsor_organization_id: str | None = None
    award_status: str | None = None
    original_start_date: datetime.date | None = None
    current_end_date: datetime.date | None = None
    adjusted_budget: float | None = None
    cumulative_funding_amount: float | None = None
    needs_veras_title: bool = False
    title_source_type: str | None = None
    model_config = {"from_attributes": True}


class DocumentRead(BaseModel):
    document_id: str
    document_type_code: str | None = None
    document_type_label: str | None = None
    parent_entity_type: str | None = None
    parent_entity_id: str | None = None
    linked_proposal_id: str | None = None
    top_level_bucket: str | None = None
    filename: str | None = None
    extension: str | None = None
    mime_type: str | None = None
    file_size_bytes: int | None = None
    storage_relative_path: str | None = None
    is_content_duplicate: bool = False
    needs_review: str | None = None
    model_config = {"from_attributes": True}


class ProjectRoleRead(BaseModel):
    project_role_id: str
    project_id: str | None = None
    personnel_id: str | None = None
    role_code: str | None = None
    key_personnel: bool = False
    funding_award_id: str | None = None
    model_config = {"from_attributes": True}


class AllowedValueRead(BaseModel):
    allowed_value_id: str
    group_name: str | None = None
    code: str | None = None
    label: str | None = None
    description: str | None = None
    model_config = {"from_attributes": True}


# ---------------------------------------------------------------------------
# Aggregated pilot dashboard schemas
# ---------------------------------------------------------------------------

class PilotRecordSummary(BaseModel):
    """Summary card for one pilot record on the dashboard."""
    record_id: str
    description: str
    title: str | None = None
    award_count: int = 0
    proposal_count: int = 0
    document_count: int = 0
    duplicate_count: int = 0
    total_budget: float = 0.0
    has_title_gaps: bool = False
    has_award: bool = True


class PilotDashboard(BaseModel):
    """Top-level dashboard response."""
    records: list[PilotRecordSummary]
    total_awards: int = 0
    total_proposals: int = 0
    total_documents: int = 0
    total_duplicates: int = 0
    total_budget: float = 0.0


class PilotRecordDetail(BaseModel):
    """Full detail view for a single pilot record."""
    record_id: str
    description: str
    awards: list[AwardRead] = []
    proposals: list[ProposalRead] = []
    projects: list[ProjectRead] = []
    documents: list[DocumentRead] = []
    personnel: list[PersonnelRead] = []
    document_count: int = 0
    duplicate_count: int = 0
    category_counts: dict[str, int] = {}


class DocumentFilterResult(BaseModel):
    """Paginated document list with facets."""
    items: list[DocumentRead]
    total_count: int
    page: int
    page_size: int
    category_counts: dict[str, int] = {}
